File: packages/septr-node/src/__tests__/secrets.test.ts
import { describe, it, expect } from "vitest"
import { detectSecrets, detectHighEntropySecrets, shouldStripKey } from "../core/secrets"

describe("detectSecrets", () => {
  it("detects OpenAI API keys", () => {
    const result = detectSecrets("sk-proj-" + "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
    expect(result.length).toBeGreaterThan(0)
    expect(result[0].type).toBe("secret_exposure")
    expect(result[0].patternId).toContain("openai")
  })

  it("detects AWS keys", () => {
    const result = detectSecrets("AKIA" + "XXXXXXXXXXXXXXXX")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects AWS secret keys", () => {
    const result = detectSecrets("wJalrXUtnFEMI/K7MDENG/" + "bPxRfiCYEXAMPLEKEY")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects Stripe live keys", () => {
    const result = detectSecrets("sk_live_" + "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects JWT tokens", () => {
    const result = detectSecrets("eyJhbGciOiJIUzI1NiJ9." + "eyJzdWIiOiIxMjM0NTY3ODkwIn0.dozjgNryP4J3jVmNHl0w5N_XgL0n3I9PlFUP0THsR8")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects private keys", () => {
    const result = detectSecrets("-----BEGIN PRIVATE" + " KEY-----\nMIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQg\n-----END PRIVATE" + " KEY-----")
    expect(result.length).toBeGreaterThan(0)
  })

  it("detects database URIs", () => {
    const result = detectSecrets("postgres://user:" + "password@localhost:5432/db")
    expect(result.length).toBeGreaterThan(0)
  })

  it("returns empty for safe input", () => {
    const result = detectSecrets("hello world this is safe")
    expect(result.length).toBe(0)
  })

  it("matches custom patterns", () => {
    const result = detectSecrets("my-secret-token-12345", ["secret-token-\\d+"])
    expect(result.length).toBeGreaterThan(0)
    expect(result[0].patternId).toBe("secret_custom")
  })

  it("handles empty string", () => {
    expect(detectSecrets("")).toEqual([])
  })

  it("configures severity correctly", () => {
    const privateKey = detectSecrets("-----BEGIN EC PRIVATE" + " KEY-----\nMIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQg\n-----END EC PRIVATE" + " KEY-----")
    expect(privateKey[0].severity).toBe("critical")

    const stripe = detectSecrets("sk_live_" + "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
    expect(stripe[0].severity).toBe("high")

    const github = detectSecrets("ghp_" + "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
    expect(github[0].severity).toBe("high")
  })
})

describe("shouldStripKey", () => {
  it("strips known sensitive keys", () => {
    expect(shouldStripKey("password")).toBe(true)
    expect(shouldStripKey("apiKey")).toBe(true)
    expect(shouldStripKey("credit_card")).toBe(true)
    expect(shouldStripKey("ssn")).toBe(true)
  })

  it("strips normalized variants", () => {
    expect(shouldStripKey("API_KEY")).toBe(true)
    expect(shouldStripKey("CreditCard")).toBe(true)
  })

  it("does not strip safe keys", () => {
    expect(shouldStripKey("name")).toBe(false)
    expect(shouldStripKey("email")).toBe(false)
    expect(shouldStripKey("description")).toBe(false)
  })

  it("strips custom fields", () => {
    expect(shouldStripKey("myCustomField", ["myCustomField"])).toBe(true)
  })

  it("handles empty key", () => {
    expect(shouldStripKey("")).toBe(false)
  })
})

describe("enhanced secret patterns", () => {
  it("detects OpenAI service account keys", () => {
    const events = detectSecrets("sk-svcacct-" + "abcdefghijklmnopqrstuvwxyz123456")
    expect(events.some((e) => e.patternId === "secret_openai_svc")).toBe(true)
  })

  it("detects GitHub fine-grained tokens", () => {
    const events = detectSecrets("github_pat_" + "11ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890")
    expect(events.some((e) => e.patternId === "secret_github_fine_grained")).toBe(true)
  })

  it("detects Twilio keys", () => {
    const events = detectSecrets("SK0123456789abcdef" + "0123456789abcdef")
    expect(events.some((e) => e.patternId === "secret_twilio")).toBe(true)
  })

  it("detects Slack webhooks", () => {
    const events = detectSecrets("https://hooks.slack.com/services/" + "T00000000/B00000000/XXXXXXXXXXXXXXXXXXXXXXXX")
    expect(events.some((e) => e.patternId === "secret_slack_webhook")).toBe(true)
  })

  it("detects Shopify tokens", () => {
    const events = detectSecrets("shpat_" + "0123456789abcdef0123456789abcdef")
    expect(events.some((e) => e.patternId === "secret_shopify")).toBe(true)
  })

  it("detects Discord bot tokens", () => {
    const events = detectSecrets("M0gQ1w2E3r4T5y6U7i8O9p0a.abcdef." + "ABCDEFGHIJKLMNOPQRSTUVWXYZA")
    expect(events.some((e) => e.patternId === "secret_discord_bot")).toBe(true)
  })

  it("detects Supabase service_role keys only with the right role", () => {
    const b64 = (d: unknown) => btoa(JSON.stringify(d)).replace(/=+$/, "")
    const header = b64({ alg: "HS256" })

    const svc = `${header}.${b64({ role: "service_role" })}.signature`
    const anon = `${header}.${b64({ role: "anon" })}.signature`

    const svcEvents = detectSecrets(svc)
    expect(svcEvents.some((e) => e.patternId === "secret_supabase_service_role")).toBe(true)
    expect(svcEvents.find((e) => e.patternId === "secret_supabase_service_role")!.severity).toBe("critical")

    const anonEvents = detectSecrets(anon)
    expect(anonEvents.some((e) => e.patternId === "secret_supabase_service_role")).toBe(false)
  })
})

describe("detectHighEntropySecrets", () => {
  it("detects high-entropy values assigned to secret-like keys", () => {
    const events = detectHighEntropySecrets('{"apiKey": "x9F2kQ7vL3pZ8nB4cD6mW1rT"}')
    expect(events.some((e) => e.patternId === "secret_high_entropy")).toBe(true)
  })

  it("skips UUID-like hex values", () => {
    const events = detectHighEntropySecrets('{"token": "3f2a9c1e-8b4d-47e6-9a2f-1c3d5e7b8a9f"}')
    expect(events).toEqual([])
  })

  it("skips values already matched by specific patterns", () => {
    const events = detectHighEntropySecrets('{"apiKey": "sk-proj-' + 'abcdefghijklmnopqrstuvwxyz1234567890"}')
    expect(events.some((e) => e.patternId === "secret_high_entropy")).toBe(false)
  })

  it("skips low-entropy and short values", () => {
    expect(detectHighEntropySecrets('{"token": "aaaaaaaaaaaaaaaaaaaaaaaaaaaa"}')).toEqual([])
    expect(detectHighEntropySecrets('{"token": "short123"}')).toEqual([])
  })
})

describe("FP suppression", () => {
  it("Google API keys are advisory-only (no redaction)", () => {
    const events = detectSecrets("AIzaSy" + "BQom12tzI-rybN7Sf-KfeL4nwm-Rf7PmI")
    expect(events.length).toBeGreaterThan(0)
    expect(events[0].patternId).toBe("secret_google_api")
    expect(events[0].severity).toBe("low")
    expect(events[0].redactable).toBe(false)
  })

  it("Supabase anon keys are advisory-only", () => {
    const b64 = (d: unknown) => btoa(JSON.stringify(d)).replace(/=+$/, "")
    const anon = `${b64({ alg: "HS256" })}.${b64({ role: "anon" })}.signature`
    const events = detectSecrets(anon)
    expect(events.some((e) => e.patternId === "secret_supabase_anon")).toBe(true)
    expect(events.find((e) => e.patternId === "secret_supabase_anon")!.redactable).toBe(false)
  })

  it("Supabase anon JWT is not flagged as generic_jwt", () => {
    const b64 = (d: unknown) => btoa(JSON.stringify(d)).replace(/=+$/, "")
    const anon = `${b64({ alg: "HS256" })}.${b64({ role: "anon" })}.signature`
    const events = detectSecrets(anon)
    expect(events.some((e) => e.patternId === "secret_generic_jwt")).toBe(false)
  })

  it("Publishable keys are not flagged by high-entropy detector", () => {
    expect(detectHighEntropySecrets('{"apiKey": "pk_live_Y2xlcmsuY2xlcmsuY29tJA"}')).toEqual([])
    expect(detectHighEntropySecrets('{"apiKey": "phc_eNuN6Ojnk9O7uWfC17z12AK85fNR0BY6IiGVy0Gfwzw"}')).toEqual([])
  })
})

File: packages/septr-go/secrets_test.go
package septr

import (
	"encoding/base64"
	"strings"
	"testing"
)

func TestDetectSecrets_OpenAI(t *testing.T) {
	result := detectSecrets("sk-proj-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
	if len(result) == 0 {
		t.Fatal("expected detection")
	}
	if result[0].PatternID != "secret_openai" {
		t.Errorf("expected secret_openai, got %s", result[0].PatternID)
	}
}

func TestDetectSecrets_StripeLive(t *testing.T) {
	result := detectSecrets("sk_live_" + "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
	if len(result) == 0 {
		t.Fatal("expected detection")
	}
}

func TestDetectSecrets_StripeTest(t *testing.T) {
	result := detectSecrets("sk_test_" + "ABCDEFGHIJKLMNOPQRSTUVWXYZabcd")
	if len(result) == 0 {
		t.Fatal("expected detection")
	}
	if len(result[0].PatternID) == 0 {
		t.Error("expected patternId")
	}
}

func TestDetectSecrets_AWSKey(t *testing.T) {
	result := detectSecrets("AKIA" + "XXXXXXXXXXXXXXXX")
	if len(result) == 0 {
		t.Fatal("expected detection")
	}
}

func TestDetectSecrets_JWT(t *testing.T) {
	result := detectSecrets("eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxMjM0NTY3ODkwIn0.dozjgNryP4J3jVmNHl0w5N_XgL0n3I9PlFUP0THsR8")
	if len(result) == 0 {
		t.Fatal("expected detection")
	}
}

func TestDetectSecrets_DatabaseURI(t *testing.T) {
	result := detectSecrets("postgres://user:password@localhost:5432/db")
	if len(result) == 0 {
		t.Fatal("expected detection")
	}
}

func TestDetectSecrets_SafeInput(t *testing.T) {
	result := detectSecrets("hello world this is safe")
	if len(result) != 0 {
		t.Fatal("expected no detection")
	}
}

func TestDetectSecrets_CustomPatterns(t *testing.T) {
	result := detectSecrets("my-secret-token-12345", []string{"secret-token-\\d+"})
	if len(result) == 0 {
		t.Fatal("expected detection")
	}
	if result[0].PatternID != "secret_custom" {
		t.Errorf("expected secret_custom, got %s", result[0].PatternID)
	}
}

func TestShouldStripKey_Empty(t *testing.T) {
	if shouldStripKey("") {
		t.Error("expected false for empty key")
	}
}

func TestShouldStripKey_Password(t *testing.T) {
	if !shouldStripKey("password") {
		t.Error("expected true for password")
	}
}

func TestShouldStripKey_Normalized(t *testing.T) {
	if !shouldStripKey("my_password_hash") {
		t.Error("expected true for my_password_hash")
	}
}

func TestShouldStripKey_CaseInsensitive(t *testing.T) {
	if !shouldStripKey("API_KEY") {
		t.Error("expected true for API_KEY")
	}
}

func TestShouldStripKey_CustomFields(t *testing.T) {
	if !shouldStripKey("custom_secret_key", []string{"custom_secret"}) {
		t.Error("expected true with custom fields")
	}
}

func TestDetectSecrets_NewPatterns(t *testing.T) {
	cases := []struct {
		id    string
		input string
	}{
		{"secret_openai_svc", "sk-svcacct-abcdefghijklmnopqrstuvwxyz123456"},
		{"secret_github_fine_grained", "github_pat_11ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"},
		{"secret_twilio", "SK0123456789abcdef" + "0123456789abcdef"},
		{"secret_slack_webhook", "https://hooks.slack.com/services/" + "T00000000/B00000000/XXXXXXXXXXXXXXXXXXXXXXXX"},
		{"secret_shopify", "shpat_" + "0123456789abcdef0123456789abcdef"},
		{"secret_discord_bot", "M0gQ1w2E3r4T5y6U7i8O9p0a.abcdef.ABCDEFGHIJKLMNOPQRSTUVWXYZA"},
		{"secret_google_client_secret", "GOCSPX-abcdefghijklmnopqrstuvwxyz123456"},
	}
	for _, c := range cases {
		events := detectSecrets(c.input)
		found := false
		for _, e := range events {
			if e.PatternID == c.id {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("expected %s for input %q", c.id, c.input)
		}
	}
}

func TestDetectSecrets_SupabaseServiceRole(t *testing.T) {
	header := base64URLEncode(`{"alg":"HS256"}`)
	svc := header + "." + base64URLEncode(`{"role":"service_role"}`) + ".signature"
	anon := header + "." + base64URLEncode(`{"role":"anon"}`) + ".signature"

	svcEvents := detectSecrets(svc)
	found := false
	for _, e := range svcEvents {
		if e.PatternID == "secret_supabase_service_role" {
			found = true
			if e.Severity != "critical" {
				t.Errorf("expected critical severity, got %s", e.Severity)
			}
		}
	}
	if !found {
		t.Error("expected service_role detection")
	}

	for _, e := range detectSecrets(anon) {
		if e.PatternID == "secret_supabase_service_role" {
			t.Error("anon key must not be flagged as service_role")
		}
	}
}

func TestDetectHighEntropySecrets(t *testing.T) {
	events := DetectHighEntropySecrets(`{"apiKey": "x9F2kQ7vL3pZ8nB4cD6mW1rT"}`)
	if len(events) == 0 || events[0].PatternID != "secret_high_entropy" {
		t.Fatalf("expected entropy detection, got %v", events)
	}

	if len(DetectHighEntropySecrets(`{"token": "3f2a9c1e-8b4d-47e6-9a2f-1c3d5e7b8a9f"}`)) != 0 {
		t.Error("UUID-like values must not be flagged")
	}
	if len(DetectHighEntropySecrets(`{"token": "aaaaaaaaaaaaaaaaaaaaaaaaaaaa"}`)) != 0 {
		t.Error("low-entropy values must not be flagged")
	}
}

func TestStripEntropyDoesNotRedact(t *testing.T) {
	value := "x9F2kQ7vL3pZ8nB4cD6mW1rT"
	note := `{"apiKey": "` + value + `"}`
	cleaned, dets := stripSensitiveData(map[string]interface{}{"note": note})
	if cleaned.(map[string]interface{})["note"] != note {
		t.Error("entropy-only value must not be redacted")
	}
	hasEntropy := false
	for _, d := range dets {
		if d.PatternID == "secret_high_entropy" {
			hasEntropy = true
		}
	}
	if !hasEntropy {
		t.Error("expected entropy detection to be reported")
	}
}

func TestDetectSecrets_SupabaseAnon_AdvisoryOnly(t *testing.T) {
	header := base64URLEncode(`{"alg":"HS256"}`)
	anon := header + "." + base64URLEncode(`{"role":"anon"}`) + ".signature"
	events := detectSecrets(anon)

	foundAnon := false
	for _, e := range events {
		if e.PatternID == "secret_supabase_anon" {
			foundAnon = true
			if e.Severity != "low" {
				t.Errorf("expected low severity for anon, got %s", e.Severity)
			}
			if e.Redactable == nil || *e.Redactable {
				t.Error("supabase_anon must be non-redactable")
			}
		}
		if e.PatternID == "secret_generic_jwt" {
			t.Error("supabase anon JWT must not be flagged as generic_jwt")
		}
	}
	if !foundAnon {
		t.Error("expected supabase_anon detection")
	}
}

func TestDetectSecrets_GoogleAPI_AdvisoryOnly(t *testing.T) {
	events := detectSecrets("AIzaSy" + "BQom12tzI-rybN7Sf-KfeL4nwm-Rf7PmI")
	found := false
	for _, e := range events {
		if e.PatternID == "secret_google_api" {
			found = true
			if e.Severity != "low" {
				t.Errorf("expected low severity for google_api, got %s", e.Severity)
			}
			if e.Redactable == nil || *e.Redactable {
				t.Error("google_api must be non-redactable")
			}
		}
	}
	if !found {
		t.Error("expected google_api detection")
	}
}

func TestDetectSecrets_SupabaseServiceRole_NotAffected(t *testing.T) {
	header := base64URLEncode(`{"alg":"HS256"}`)
	svc := header + "." + base64URLEncode(`{"role":"service_role"}`) + ".signature"
	events := detectSecrets(svc)
	found := false
	for _, e := range events {
		if e.PatternID == "secret_supabase_service_role" {
			found = true
			if e.Severity != "critical" {
				t.Errorf("expected critical severity, got %s", e.Severity)
			}
			if e.Redactable != nil && !*e.Redactable {
				t.Error("supabase_service_role must be redactable")
			}
		}
	}
	if !found {
		t.Error("expected supabase_service_role detection")
	}
}

func TestStrip_SupabaseAnon_NotRedacted(t *testing.T) {
	header := base64URLEncode(`{"alg":"HS256"}`)
	anon := header + "." + base64URLEncode(`{"role":"anon"}`) + ".signature"
	cleaned, _ := stripSensitiveData(map[string]interface{}{"data": anon})
	if cleaned.(map[string]interface{})["data"] != anon {
		t.Error("supabase anon key must not be redacted in strip")
	}
}

func TestDetectHighEntropy_PublishableKey(t *testing.T) {
	tests := []string{
		`{"apiKey": "pk_live_abc123def456ghi789"}`,
		`{"apiKey": "pk_test_abc123def456ghi789"}`,
		`{"apiKey": "pk_prod_abc123def456ghi789"}`,
		`{"apiKey": "pk.abc123def456ghi789jkl012mno"}`,
		`{"apiKey": "phc_abc123def456ghi789jkl012mno"}`,
		`{"apiKey": "phx_abc123def456ghi789jkl012mno"}`,
	}
	for _, input := range tests {
		if len(DetectHighEntropySecrets(input)) != 0 {
			t.Errorf("publishable key must not be flagged: %s", input)
		}
	}
}

func base64URLEncode(s string) string {
	encoded := base64.StdEncoding.EncodeToString([]byte(s))
	encoded = strings.ReplaceAll(encoded, "+", "-")
	encoded = strings.ReplaceAll(encoded, "/", "_")
	return strings.TrimRight(encoded, "=")
}
